packages/studio/src/store/slices/pages/createPageSlice.ts

```diff
@@ -158,6 +163,108 @@ export const createPageSlice: SliceCreator<PageSlice> = (set, get) => {
+      const activePageState = getActivePageState();
+      const componentTree = activePageState?.componentTree;
+      if (!componentTree) {
+        throw new Error(
+          `Error rendering components: component tree is undefined.`
+        );
+      }
+      const orderedSelectedComponentUUIDs =
+        ComponentTreeHelpers.mapComponentTreeParentsFirst(
+          componentTree,
+          (c) => {
+            if (selectedComponentUUIDs.has(c.uuid)) {
+              return c.uuid;
+            }
+            return undefined;
+          }
+        ).filter((uuid) => !!uuid);
+      return orderedSelectedComponentUUIDs.length !== 0
+        ? orderedSelectedComponentUUIDs[0]
+        : undefined;
+    },
+    addSelectedComponentUUID: (selectedUUID: string) => {
+      set((store) => {
+        store.selectedComponentUUIDs.add(selectedUUID);
+      });
+    },
+    addSelectedComponentRect: (
+      selectedUUID: string,
+      rect: DOMRectProperties
+    ) => {
+      set((store) => {
+        store.selectedComponentRectsMap[selectedUUID] = rect;
+      });
+    },
+    clearSelectedComponents: () => {
+      set((store) => {
+        store.selectedComponentUUIDs = new Set();
+        store.selectedComponentRectsMap = {};
+      });
+    },
+    /**
+     * Adds all of the components and their children ordered between the selected component and
+     * the active component (inclusive) to selectedComponentUUIDs.
+     */
+    addShiftSelectedComponentUUIDs: (selectedComponent: ComponentState) => {
+      const selectedUUID = selectedComponent.uuid;
+      const {
+        activeComponentUUID,
+        getActivePageState,
+        addSelectedComponentUUID,
+        clearSelectedComponents,
+      } = get();
+      const activePageState = getActivePageState();
+      const componentTree = activePageState?.componentTree;
+      if (selectedUUID === activeComponentUUID) {
+        clearSelectedComponents();
+        addSelectedComponentUUID(selectedUUID);
+        return;
+      }
+      if (!componentTree) {
+        throw new Error(
+          `Error selecting components: component tree is undefined.`
+        );
+      }
+      if (!activeComponentUUID) {
+        throw new Error(
+          `Error selecting components: active component is undefined.`
+        );
+      }
+      const targetComponentUUIDs = [selectedUUID, activeComponentUUID] as const;
+      const lowestParentUUID = ComponentTreeHelpers.getLowestParentUUID(
+        ...targetComponentUUIDs,
+        componentTree
+      );
+
+      let selecting = false;
+      ComponentTreeHelpers.mapComponentTreeParentsFirst(componentTree, (c) => {
+        if (c.parentUUID !== lowestParentUUID) return;
+        const descendants = ComponentTreeHelpers.getDescendants(
+          c,
+          componentTree
+        );
+        const targetIsParent = targetComponentUUIDs.includes(c.uuid);
+        const targetInDescendants = descendants.some((d) =>
```

Comment:
I still find this logic a little complicated. We don't actually care where the target is present as long as it's in the sub-tree that starts with the current component or not. Could we update the callback so we get the sub-tree (not just descendants) and then use `reduce` to get the number of targets present in the sub-tree? then the handling logic could be simpler, something like
```ts
if (numTargetsInSubTree === 0 && !selecting) {
  return;
}
subTree.forEach((d) => addSelectedComponentUUID(d.uuid));
if (numTargetsInSubTree === 1) {
  selecting = !selecting;
}
```

Reply:
Done. Rather than counting the number of targets in the subtree, I'm checking if either of the targets is in the subtree at all, and I left the check at the beginning that returns if the current parent UUID is not the lowestParentUUID. I think it's a little easier to follow this way but I'm happy to change it if you disagree.

Reply:
We still need to differentiate between the case where both targets are in the subtree vs. only one. We wouldn't want to start selecting if both targets are in the current subtree because then we'd never stop selecting

Reply:
I believe this is handled by the `if (c.parentUUID !== lowestParentUUID) return;` check at the beginning. If the parent is the lowestParent, and both targets are in the subtree, then the one must be a child of the other, in which case I believe we would want to select all children of the higher component. I'll switch it to count the number of targets in the subTree since that lets us get rid of lowestParentUUID entirely, but I _think_ both should work.

Reply:
I tried simplifying the logic further but all of the very simple solutions had issues with various edge cases. I updated it to handle all the cases I could find, and the flow is roughly as follows:

- `selecting` starts as `false` and only changes when the selected or active components are reached (I think it might have been swapping back and force multiple times previously)
- If the current parent is not the lowest common parent between the selected and active components, return; we never want to select anything higher up in the tree than lowest common parent
- Generate subtree
- If either the subtree contains one of the targets or `selecting` is `true`, then add all elements in the subtree to `selectedComponentUUIDs`